File: generator/audio_check_test.py
"""Test für audio_check.py - die Audio/Skript-Tempo-Plausibilitätsprüfung.

Läuft komplett OHNE ffmpeg und ohne echtes Video: die Tempo-Schätzung und der
Vergleich sind reine Funktionen auf synthetischen Signalen (Sinuskurven
bekannter Frequenz statt echter Aufnahmen) - genau wie quality_doctor_test.py
seine Rhythmusprüfung an erzeugten, nicht echten Kurven testet.

Ausführen: python3 generator/audio_check_test.py
"""

import sys

import numpy as np

import audio_check as ac


def sine(freq_hz, duration_s, sample_rate, amplitude=1.0, phase=0.0):
    t = np.arange(0, duration_s, 1.0 / sample_rate)
    return amplitude * np.sin(2 * np.pi * freq_hz * t + phase)


def amplitude_modulated_noise(envelope_freq_hz, duration_s, sample_rate, rng):
    """Simuliert eine Tonspur, deren LAUTSTÄRKE mit envelope_freq_hz
    pulsiert (wie rhythmisches Stöhnen/Geräusch), aber deren Trägersignal
    selbst hochfrequentes Rauschen ist - realistischer als eine reine
    Sinuswelle für "Audioenergie", ohne eine echte Aufnahme zu brauchen."""
    t = np.arange(0, duration_s, 1.0 / sample_rate)
    envelope = 0.5 + 0.5 * np.sin(2 * np.pi * envelope_freq_hz * t)
    carrier = rng.normal(0, 1, size=len(t))
    return envelope * carrier


def script_actions(freq_hz, duration_ms, step_ms=40):
    at = np.arange(0, duration_ms, step_ms, dtype=float)
    pos = 50 + 45 * np.sin(2 * np.pi * freq_hz * at / 1000.0)
    return [{"at": int(a), "pos": float(p)} for a, p in zip(at, pos)]


def main():
    failures = []

    def check(name, cond, detail=""):
        print(("  OK   " if cond else "  FAIL ") + name + (f"  [{detail}]" if not cond else ""))
        if not cond:
            failures.append(name)

    # --- _dominant_frequency_hz: erkennt eine bekannte Sinusfrequenz -------
    sr = 100.0
    sig = sine(1.5, 20, sr)
    freq = ac._dominant_frequency_hz(sig, sr)
    check("erkennt 1.5Hz in einem reinen Sinussignal",
          freq is not None and abs(freq - 1.5) < 0.1, str(freq))

    # --- zu kurzes Signal -> None statt Absturz ------------------------------
    check("zu kurzes Signal liefert None", ac._dominant_frequency_hz([1, 2, 3], sr) is None, "")

    # --- audio_envelope: RMS-Hüllkurve einer amplitudenmodulierten Spur -----
    rng = np.random.default_rng(0)
    audio_sr = 4000
    samples = amplitude_modulated_noise(1.0, 30, audio_sr, rng)
    envelope, env_rate = ac.audio_envelope(samples, audio_sr, window_ms=50.0)
    check("Hüllkurve ist deutlich kürzer als das Rohsignal",
          0 < len(envelope) < len(samples), f"{len(envelope)} vs {len(samples)}")
    check("Hüllkurvenrate entspricht 1000/window_ms", env_rate == 20.0, str(env_rate))

    # --- estimate_audio_tempo_hz: erkennt die Hüllkurven-Frequenz -----------
    audio_hz = ac.estimate_audio_tempo_hz(samples, audio_sr, window_ms=50.0)
    check("erkennt 1.0Hz-Lautstärkepulsierung", audio_hz is not None and abs(audio_hz - 1.0) < 0.15,
          str(audio_hz))

    # --- estimate_script_tempo_hz: erkennt die Skript-eigene Frequenz -------
    actions = script_actions(1.0, 20000)
    script_hz = ac.estimate_script_tempo_hz(actions)
    check("erkennt 1.0Hz im Skript", script_hz is not None and abs(script_hz - 1.0) < 0.15,
          str(script_hz))

    check("zu wenige Actions liefert None", ac.estimate_script_tempo_hz(actions[:3]) is None, "")

    # --- compare_tempo: exakte Übereinstimmung -------------------------------
    cmp_exact = ac.compare_tempo(1.0, 1.0)
    check("1:1 Übereinstimmung erkannt", cmp_exact["matches"] and cmp_exact["harmonic"] == 1.0,
          str(cmp_exact))

    # --- compare_tempo: doppeltes Audio-Tempo (2 Töne pro Hub) --------------
    cmp_double = ac.compare_tempo(1.0, 2.0)
    check("erkennt 0.5x als passendes Vielfaches (Audio doppelt so schnell)",
          cmp_double["matches"] and cmp_double["harmonic"] == 0.5, str(cmp_double))

    # --- compare_tempo: klarer Ausreißer, kein Vielfaches passt --------------
    cmp_off = ac.compare_tempo(1.0, 3.7)
    check("deutliche Abweichung wird NICHT als Übereinstimmung gewertet",
          not cmp_off["matches"], str(cmp_off))

    # --- compare_tempo: fehlende Werte -> None statt falscher Vergleich -----
    check("None ohne script_hz", ac.compare_tempo(None, 1.0) is None, "")
    check("None ohne audio_hz", ac.compare_tempo(1.0, None) is None, "")

    # --- extract_audio_samples: klarer Fehler ohne ffmpeg/Datei -------------
    raised = False
    try:
        ac.extract_audio_samples("/nicht/vorhanden.mp4")
    except ac.AudioUnavailable:
        raised = True
    check("extract_audio_samples wirft AudioUnavailable statt abzustürzen", raised, "")

    # --- check(): Gesamtablauf ohne echtes Video -> ehrliche Nichtverfügbarkeit
    result = ac.check("/nicht/vorhanden.mp4", actions)
    check("check() meldet available=False ohne echtes Video, statt zu crashen",
          result["available"] is False and result["warnings"] == [], str(result))

    print(("FEHLGESCHLAGEN: " + ", ".join(failures)) if failures else "Alle Prüfungen bestanden.")
    return 1 if failures else 0


if __name__ == "__main__":
    sys.exit(main())

File: generator/audiocheck_test.go
package generator

import (
	"math"
	"testing"

	"github.com/funfunpayer/SamNPlayer/funscript"
)

func sine(freqHz, durationS, sampleRate float64) []float64 {
	n := int(durationS * sampleRate)
	out := make([]float64, n)
	for i := 0; i < n; i++ {
		out[i] = math.Sin(2 * math.Pi * freqHz * float64(i) / sampleRate)
	}
	return out
}

func TestDominantFrequencyHz(t *testing.T) {
	sr := 100.0
	sig := sine(1.5, 20, sr)
	freq := dominantFrequencyHz(sig, sr, audioMinTempoHz, audioMaxTempoHz)
	if freq == nil || math.Abs(*freq-1.5) > 0.1 {
		t.Fatalf("expected ~1.5Hz, got %v", freq)
	}
	if dominantFrequencyHz([]float64{1, 2, 3}, sr, audioMinTempoHz, audioMaxTempoHz) != nil {
		t.Fatal("short signal must yield nil")
	}
}

func TestEstimateScriptTempoHz(t *testing.T) {
	actions := make([]funscript.Action, 0, 500)
	for at := int64(0); at < 20000; at += 40 {
		pos := 50 + 45*math.Sin(2*math.Pi*1.0*float64(at)/1000.0)
		actions = append(actions, funscript.Action{At: at, Pos: int(math.Round(pos))})
	}
	hz := estimateScriptTempoHz(actions)
	if hz == nil || math.Abs(*hz-1.0) > 0.15 {
		t.Fatalf("expected ~1.0Hz, got %v", hz)
	}
	if estimateScriptTempoHz(actions[:3]) != nil {
		t.Fatal("too few actions must yield nil")
	}
}

func TestAppendQualityAudioHint(t *testing.T) {
	hz := 1.25
	audio := &funscript.AudioCheck{AudioHz: &hz}
	AppendQualityAudioHint(funscript.ScriptQualityResult{Passed: true, Score: 0.9}, audio)
	if len(audio.Warnings) != 0 {
		t.Fatalf("passed quality must not add hint, got %v", audio.Warnings)
	}
	AppendQualityAudioHint(funscript.ScriptQualityResult{Passed: false, Score: 0.2}, audio)
	if len(audio.Warnings) != 1 {
		t.Fatalf("want 1 warning, got %v", audio.Warnings)
	}
	AppendQualityAudioHint(funscript.ScriptQualityResult{Passed: false, Score: 0.2}, audio)
	if len(audio.Warnings) != 1 {
		t.Fatalf("hint must be idempotent, got %v", audio.Warnings)
	}
	noHz := &funscript.AudioCheck{}
	AppendQualityAudioHint(funscript.ScriptQualityResult{Passed: false}, noHz)
	if len(noHz.Warnings) != 0 {
		t.Fatal("no AudioHz → no hint")
	}
}

func TestCompareTempo(t *testing.T) {
	one, two, off := 1.0, 2.0, 3.7
	exact := compareTempo(&one, &one, 0.25, audioDefaultHarmonics)
	if exact == nil || !exact.matches || exact.harmonic != 1.0 {
		t.Fatalf("1:1: %+v", exact)
	}
	dbl := compareTempo(&one, &two, 0.25, audioDefaultHarmonics)
	if dbl == nil || !dbl.matches || dbl.harmonic != 0.5 {
		t.Fatalf("0.5x: %+v", dbl)
	}
	bad := compareTempo(&one, &off, 0.25, audioDefaultHarmonics)
	if bad == nil || bad.matches {
		t.Fatalf("outlier should not match: %+v", bad)
	}
	if compareTempo(nil, &one, 0.25, audioDefaultHarmonics) != nil {
		t.Fatal("nil script_hz")
	}
}

func TestAudioEnvelope(t *testing.T) {
	sr := 4000.0
	samples := make([]float64, int(30*sr))
	for i := range samples {
		env := 0.5 + 0.5*math.Sin(2*math.Pi*1.0*float64(i)/sr)
		// deterministic "noise" carrier
		carrier := math.Sin(2*math.Pi*200*float64(i)/sr) + 0.3*math.Sin(2*math.Pi*350*float64(i)/sr)
		samples[i] = env * carrier
	}
	env, rate := audioEnvelope(samples, sr, 50.0)
	if len(env) < 8 || rate != 20.0 {
		t.Fatalf("envelope len=%d rate=%v", len(env), rate)
	}
	hz := estimateAudioTempoHz(samples, sr, 50.0)
	if hz == nil || math.Abs(*hz-1.0) > 0.2 {
		t.Fatalf("expected ~1.0Hz audio tempo, got %v", hz)
	}
}

func TestCheckAudioTempoNoAudio(t *testing.T) {
	if CheckAudioTempo("/nonexistent/nope.mp4", []funscript.Action{{At: 0, Pos: 0}}) != nil {
		t.Fatal("missing video must return nil")
	}
}
